### PR TITLE
Change Date to number of days after Unix epoch

### LINKED ISSUE #2803: Change Date representation to an int64 Time in seconds representation
Currently Date is represented as a time.Time. While this is okay, we've run into a number of problems converting Date into Time, or vice versa, mostly as a result of the fallout of having a time zone offset in time.Time. It will be safer and better to just represent date as an int64 UTC Unix time in seconds.

### PR DESCRIPTION
**Pull request by vivekmenezes**
closes #2803

### CODE REVIEW

File: sql/parser/datum.go [hunk 1]
```diff
@@ -364,23 +364,25 @@ func (d DDate) Compare(other Datum) int {
...
 
 // IsMin implements the Datum interface.
 func (d DDate) IsMin() bool {
-	// Subtracting a day underflows to a larger value.
-	return d.Before(d.AddDate(0, 0, -1))
+	return d == math.MinInt64
+}
+
+func (d DDate) toTime() time.Time {
+	return time.Unix(int64(d)*driver.SecondsInDay, 0).UTC()
```
Comment (vivekmenezes) [inline sql/parser/datum.go]: Can overflow but time.Unix() itself can overflow for a large int64.

File: sql/parser/eval.go [hunk 2]
```diff
@@ -440,7 +441,7 @@ var cmpOps = map[cmpArgs]cmpOp{
 	},
 	cmpArgs{LE, dateType, dateType}: {
 		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
-			return DBool(right.(DDate).Before(left.(DDate).Time)), nil
```
Comment (vivekmenezes) [inline sql/parser/eval.go]: This was a bug

Comment (tschottdorf) [inline sql/parser/eval.go]: could you add a test case that now fails?

File: sql/parser/eval.go [hunk 3]
```diff
@@ -567,13 +568,14 @@ var defaultContext = EvalContext{
...
+func (ctx EvalContext) makeDDate(t time.Time) (DDate, error) {
 	loc, err := ctx.GetLocation()
 	if err != nil {
-		return DNull, err
+		return DummyDate, err
 	}
 	year, month, day := t.In(loc).Date()
-	return DDate{Time: time.Date(year, month, day, 0, 0, 0, 0, time.UTC)}, nil
+	secs := time.Date(year, month, day, 0, 0, 0, 0, time.UTC).Unix()
+	return DDate(secs / driver.SecondsInDay), nil
```
Comment (petermattis) [inline sql/parser/eval.go]: I think not repeating yourself by using the `driver.SecondsInDay` constant is not worth the dependency on the `sql/driver` package. I'd probably feel differently if the constant could change, but `SecondsInDay` is a constant in the true sense: it can never have a value other than `24 * 60 * 60`.

**Comment by petermattis** [comment]
LGTM

File: sql/parser/eval.go (hunk 2 above)
Comment (vivekmenezes) [inline sql/parser/eval.go]: done

File: sql/parser/eval.go (hunk 3 above)
Comment (vivekmenezes) [inline sql/parser/eval.go]: done

**Comment by petermattis** [comment]
LGTM (again)

File: sql/driver/wire.go [hunk 4]
```diff
@@ -101,7 +102,8 @@ func (d Datum) Value() (driver.Value, error) {
 	case *Datum_StringVal:
 		val = t.StringVal
 	case *Datum_DateVal:
-		val = MakeDate(t.DateVal.GoTime())
+		timestamp := time.Unix(t.DateVal*secondsInDay, 0).UTC()
+		val = MakeDate(timestamp)
```
Comment (tamird) [inline sql/driver/wire.go]: can you also change `driver.Date` to the same representation as parser.DDate?

**Comment by tamird** [comment]
We need to apply this change to `driver.Date` as well as `parser.DDate`

**Comment by vivekmenezes** [comment]
will do

File: sql/parser/eval.go [hunk 5]
```diff
@@ -136,13 +139,13 @@ var binOps = map[binArgs]binOp{
 	binArgs{Plus, dateType, intervalType}: {
 		returnType: DummyTimestamp,
 		fn: func(left Datum, right Datum) (Datum, error) {
-			return DTimestamp{Time: left.(DDate).Add(right.(DInterval).Duration)}, nil
+			return DTimestamp{Time: left.(DDate).toTime().Add(right.(DInterval).Duration)}, nil
```
Comment (bdarnell) [inline sql/parser/eval.go]: `toTime` always uses UTC; a date plus or minus an interval or timestamp should be computed in the session time zone. (and we need some tests that expose this)

`toTime` is rarely useful and it looks like we only ever use it correctly when we want to pull the (year, month, day) fields out of a value so we should probably replace `toTime()` with a `toDate() (int, int, int)` method.

Comment (vivekmenezes) [inline sql/parser/eval.go]: I think we need to remove this type of casting of date to time in the
binary operations. I'm going to remove them.

Comment (vivekmenezes) [inline sql/parser/eval.go]: I've taken out a few binary operations like timestamp +- date. Ready for another viewing. Thanks

File: sql/parser/eval.go [hunk 6]
```diff
@@ -134,15 +137,23 @@ var binOps = map[binArgs]binOp{
...
 	binArgs{Plus, dateType, intervalType}: {
-		returnType: DummyTimestamp,
+		returnType: DummyDate,
 		fn: func(left Datum, right Datum) (Datum, error) {
-			return DTimestamp{Time: left.(DDate).Add(right.(DInterval).Duration)}, nil
+			days, err := inDays(right.(DInterval).Duration)
+			if err != nil {
+				return DummyDate, err
+			}
+			return DDate(left.(DDate) + days), nil
```
Comment (petermattis) [inline sql/parser/eval.go]: I'm not clear on why this change was made. The previous implementation matched the postgres behavior. This one does not.

File: sql/parser/eval.go [hunk 7]
```diff
@@ -316,6 +319,13 @@ var binOps = map[binArgs]binOp{
 	},
 }
 
+func inDays(d time.Duration) (DDate, error) {
+	if d%(24*time.Hour) != 0 {
```
Comment (bdarnell) [inline sql/parser/eval.go]: s/24*time.Hour/time.Day/g

File: sql/parser/eval.go (hunk 6 above)
Comment (vivekmenezes) [inline sql/parser/eval.go]: if you want the postgres behavior, you have to explicitly cast the date to a timestamp and then add an interval. We basically only allow dates to be modified in days specified as in interval. Another option is to forget about the whole interval crap and use an integer for days which is what postgres does.

File: sql/parser/eval.go (hunk 7 above)
Comment (vivekmenezes) [inline sql/parser/eval.go]: There is no such things as time.Day.

File: sql/parser/eval.go (hunk 6 above)
Comment (petermattis) [inline sql/parser/eval.go]: Not sure what you mean about forgetting the interval stuff. Certainly looks like postgres has an interval type and that it has sub-second granularity.

File: sql/parser/eval.go (hunk 7 above)
Comment (bdarnell) [inline sql/parser/eval.go]: Ah, right. Because days are not always 24 hours (daylight savings). Which means that the result of `inDays` cannot be used in an operation on any time that is not already in UTC. It would be better to make this `addDays(time.Time, time.Duration)` which would either assert that the given time is in UTC or do the right thing in the given time's time zone.

File: sql/parser/eval.go (hunk 6 above)
Comment (bdarnell) [inline sql/parser/eval.go]: @vivekmenezes's point (with this change; I also don't understand "forget about intervals") is that adding an interval to a `DATE` is a kind of implicit cast (unless that interval is an integer number of days), just like adding an int and a float. It would be more consistent with our stance on implicit casts elsewhere to require the date to be explicitly cast to a timestamp (even though it is inconsistent with postgres)

Comment (petermattis) [inline sql/parser/eval.go]: Ah, that makes more sense, though (sadly) I believe we're going to have to revisit our stance on implicit casts as no other system is as strict as we are.

Comment (bdarnell) [inline sql/parser/eval.go]: I think that remains to be seen. It's probably true that no other system is as strict as we are, but how much does that matter? Most ORMs are pretty conservative in the SQL they generate and would be compatible with a strict server-side typesystem. The place where strictness gets annoying is the interactive console, which is has such different requirements that I wonder if we should offer a special non-strict mode for it.

Comment (vivekmenezes) [inline sql/parser/eval.go]: I've modified it to only support operations between dates, and between a date and an integer. I've taken out arithmetic between date and timestamp and between date and interval. A user will need to explicitly cast a date to a timestamp to do that.

File: sql/parser/eval.go (hunk 7 above)
Comment (vivekmenezes) [inline sql/parser/eval.go]: I've taken out support for arithmetic between a date and a duration and supplanted it with arithmetic between a date and an int.

File: sql/parser/datum.go [hunk 8]
```diff
@@ -333,10 +333,9 @@ func (d DBytes) String() string {
 	return encodeSQLBytes(string(d))
 }
 
-// DDate is the date Datum.
-type DDate struct {
-	time.Time // Must always be UTC!
-}
+// DDate is the date Datum represented as the number of days after
+// the Unix epoch.
+type DDate int64
```
Comment (tamird) [inline sql/parser/datum.go]: possibly pedantic: do we want to make this a struct containing an int64 so that it can't just be cast?

**Comment by tamird** [comment]
can we break out the casting changes into a separate change?

File: sql/parser/datum.go (hunk 8 above)
Comment (vivekmenezes) [inline sql/parser/datum.go]: A struct is valuable if you need to embed an interface. Adding the int64 to
a struct will make the code uglier. Luckily in Go you have to explicitly
cast so it's safe from implicit madness. I don't think this is an
improvement.

Comment (vivekmenezes) [inline sql/parser/datum.go]: I can do that

**Comment by vivekmenezes** [comment]
Moved this PR into four commits.

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
LGTM
